include/crocoddyl/multibody/costs/centroidal-momentum.hxx

```diff
@@ -38,49 +39,43 @@ CostModelCentroidalMomentumTpl<Scalar>::CostModelCentroidalMomentumTpl(
-    : Base(state, 6), href_(href), pin_model_(state->get_pinocchio()) {}
+    : Base(state, boost::make_shared<ResidualModelCentroidalMomentum>(state, href)), href_(href) {}
 
 template <typename Scalar>
 CostModelCentroidalMomentumTpl<Scalar>::~CostModelCentroidalMomentumTpl() {}
 
 template <typename Scalar>
 void CostModelCentroidalMomentumTpl<Scalar>::calc(const boost::shared_ptr<CostDataAbstract>& data,
-                                                  const Eigen::Ref<const VectorXs>&,
-                                                  const Eigen::Ref<const VectorXs>&) {
-  // Compute the cost residual give the reference CentroidalMomentum
+                                                  const Eigen::Ref<const VectorXs>& x,
+                                                  const Eigen::Ref<const VectorXs>& u) {
+  // Compute the cost residual given the reference centroidal momentum
   Data* d = static_cast<Data*>(data.get());
-  data->r = d->pinocchio->hg.toVector() - href_;
+  residual_->calc(d->residual, x, u);
 
-  activation_->calc(data->activation, data->r);
+  // Compute the cost
+  activation_->calc(data->activation, data->residual->r);
   data->cost = data->activation->a_value;
 }
 
 template <typename Scalar>
 void CostModelCentroidalMomentumTpl<Scalar>::calcDiff(const boost::shared_ptr<CostDataAbstract>& data,
-                                                      const Eigen::Ref<const VectorXs>&,
-                                                      const Eigen::Ref<const VectorXs>&) {
+                                                      const Eigen::Ref<const VectorXs>& x,
+                                                      const Eigen::Ref<const VectorXs>& u) {
+  // Compute the derivatives of the activation and centroidal momentum residual models
   Data* d = static_cast<Data*>(data.get());
-  const std::size_t& nv = state_->get_nv();
-  Eigen::Ref<Matrix6xs> Rq = data->Rx.leftCols(nv);
-  Eigen::Ref<Matrix6xs> Rv = data->Rx.rightCols(nv);
-
-  activation_->calcDiff(data->activation, data->r);
-  pinocchio::getCentroidalDynamicsDerivatives(*pin_model_.get(), *d->pinocchio, Rq, d->dhd_dq, d->dhd_dv, Rv);
-
-  // The derivative computation in pinocchio does not take the frame of reference into
-  // account. So we need to update the com frame as well.
-  for (int i = 0; i < d->pinocchio->Jcom.cols(); ++i) {
-    data->Rx.template block<3, 1>(3, i) -= d->pinocchio->Jcom.col(i).cross(d->pinocchio->hg.linear());
-  }
-
-  d->Arr_Rx.noalias() = data->activation->Arr * data->Rx;
-  data->Lx.noalias() = data->Rx.transpose() * data->activation->Ar;
-  data->Lxx.noalias() = data->Rx.transpose() * d->Arr_Rx;
+  activation_->calcDiff(data->activation, data->residual->r);
+  residual_->calcDiff(d->residual, x, u);
+
+  // Compute the derivatives of the cost function based on a Gauss-Newton approximation
+  Eigen::Ref<Matrix6xs> Rx(data->residual->Rx);
```

Comment:
The idea was to perform the above operations with a fixed number of rows.
I though this was more efficient, isn't it?

As far I understand, this code doesn't allocate data.

**Let me know if you still advice to do not use this instance**

Reply:
No, it is more efficient to use a block reference

Reply:
Could you clarify more your suggestion?

I don't understand why you suggested to use block reference.

If we don't want to use `Eigen::Ref<Matrix6xs>`, then we do need to use `data->residual->Rx` instead of `Rx` in the above multiplication. Is this what do you suggest?

Reply:
Do you have compilation issues with this?

Reply:
done